### PR TITLE
Solution react_todo-app-with-api

### PR DESCRIPTION
**Pull request by NazariiAlieksieiev**
[DEMO LINK](https://NazariiAlieksieiev.github.io/react_todo-app-with-api/)

Should I rewrite this code using useContext?

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,249 @@
...
+      .then((fetchedTodos: Todo[]) => {
+        setTodos(fetchedTodos);
+      })
+      .catch(() => {
+        setError(TodoErrors.Get);
+      });
+  }, []);
+
+  useEffect(() => {
+    let timeout: NodeJS.Timeout;
```
Comment (mykhalenych) [inline src/App.tsx:171]: you work in browser, not in node.js environment

File: src/App.tsx [hunk 2]
```diff
@@ -1,24 +1,249 @@
...
+          setSeletedId(null);
+        })
+        .catch(() => setError(TodoErrors.Update));
+    }
+  };
+
+  const itemsLeft = todos.filter(todo => !todo.completed).length;
+  const itemsCompleted = todos.filter(todo => todo.completed).length;
+
+  useEffect(() => {
```
Comment (mykhalenych) [inline src/App.tsx:160]: move useEffects to bottom before 196 line

File: src/api/todos.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  });
+};
+
+export const deleteTodo = (todoId: number) => {
+  return client.delete(`/todos/${todoId}`);
+};
+
+export const toggleTodo = (todoId: number, data: boolean) => {
+  return client.patch(`/todos/${todoId}`, {
+    completed: !data,
```
Comment (mykhalenych) [inline src/api/todos.ts:24]:
```suggestion
    completed,
```

**Review by mykhalenych — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 4]
```diff
@@ -1,24 +1,251 @@
...
+                ? { ...todo, title: newTitle }
+                : todo
+            )),
+          );
+        })
+        .finally(() => {
+          setIsProcessed(false);
+          setNewTitle('');
+          setEditedTodoId(null);
+          setSeletedId(null);
```
Comment (Viktor-Kost) [inline src/App.tsx:153]: move this code to a reset function 
```
const resetTodoData = () :void => {
  setIsProcessed(false);
  setNewTitle('');
  setEditedTodoId(null);
  setSeletedId(null);
}
```

File: src/App.tsx [hunk 5]
```diff
@@ -1,24 +1,251 @@
...
+          .filter(filtrTodo => !filtrTodo.completed)))
+        .catch(() => setError(TodoErrors.Delete)));
+  };
+
+  const changeTodoTitleHandler = (
+    event: FormEvent<HTMLFormElement> | FocusEvent<HTMLInputElement>,
+  ) => {
+    if (event) {
+      event.preventDefault();
+    }
```
Comment (Viktor-Kost) [inline src/App.tsx:119]:
```suggestion
      event.preventDefault();
```

File: src/App.tsx [hunk 6]
```diff
@@ -1,24 +1,251 @@
...
+    setSeletedId(todoId);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(filtredTodos => filtredTodos
+          .filter(todo => todo.id !== todoId));
+      })
+      .catch(() => setError(TodoErrors.Delete))
+      .finally(() => setSeletedId(null));
+  }, []);
```
Comment (Viktor-Kost) [inline src/App.tsx:40]: it'd a good practice to use async/await syntax and apply for similar methods

```
const deleteTodoHandler = useCallback(async (todoId: number) => {
  setSeletedId(todoId);

  try {
    await deleteTodo(todoId);
    setTodos(filtredTodos => filtredTodos.filter(todo => todo.id !== todoId));
  } catch (error) {
    setError(TodoErrors.Delete);
  } finally {
    setSeletedId(null);
  }
}, []);
```

**Review by Viktor-Kost — CHANGES_REQUESTED** [review]
Hello!)
Good job!

**To improve:**
1) check your calculation when todos are completed
Screenshot 2023-05-08 at 20 08 38

2) left a few comments

**Review by etojeDenys — APPROVED** [review]
LGTM

**Pull request closed without merging**